### PR TITLE
Publish heading

### PR DESCRIPTION
**Pull request by mr337**
This PR is to add support for publishing heading from GPHDT to /heading topic as a QuanternionStamped.

### CODE REVIEW

**Comment by danielsnider** [comment]
I would love to use this in mainline. Why hasn't this been merged?

**Comment by mr337** [comment]
Not for sure, there is one little fix I though I pushed I need to fix here.

https://github.com/ros-drivers/nmea_navsat_driver/pull/25/files#diff-252e39a40920f294c2bc5b12a27eaaefL127

**Comment by danielsnider** [comment]
Thanks for the heads up!!

**Comment by mr337** [comment]
@danielsnider fixed

In general not for sure what I need to do to get this reviewed.

**Comment by naoki-mizuno** [comment]
@mr337 Great work! I've tested this with our Hemisphere ssV-102 GPS compass and confirmed that it works. I wonder if @mikepurvis or @ericperko can check and merge this PR?

**Comment by mikepurvis** [comment]
I'm not actually an official maintainer of this package, but the change mostly looks reasonable to me. Some thoughts:

- Don't bump the `<version>` elements; that's the maintainer's job to do during the release process.
- What's the relationship between GPHDT and GPVTG? If GPHDT is vehicle heading rather than track heading, then presumably it's only available on a dual antenna system or a system that's fusing data from a magnetometer? Either way, let's see some comments in the source explaining this.
- Is there an end to end example of how this data would be consumed? In particular, [robot_localization isn't able to consume a QuaternionStamped](http://docs.ros.org/lunar/api/robot_localization/html/preparing_sensor_data.html), so perhaps it would be of greater value to supply a partially filled [Imu message](http://docs.ros.org/api/sensor_msgs/html/msg/Imu.html) rather than a bare quaternion?
- I haven't looked much into this, but it would be great to have a simulation story, too. Are there existing Gazebo plugins which publish this kind of sensor data? If so, what form does it take? Is there an existing API that can be followed?

**Comment by mr337** [comment]
Sorry for late reply, been on vaca.

- Let me revert the version elements, didn't realize that.
- That is determined by the system that is generating the NMEA lines. In my case, this is generated from a dual antenna setup and is only the true heading. I can throw in some comments about this.
- I've been on the fence about this, and could use a little guidance. IIRC the original reason why I went with a  QuaternionStamped is my 9DOF used that format, so this was a drop in replacement. Granted, this worked well for my needs, but isn't a good standard so I'm pretty open to what a better msg format would be.
- Tying back to the previous answer, I do use a gazebo plugin (http://wiki.ros.org/hector_gazebo_plugins#GazeboRosMagnetic), which is yet in another msg format. Quick look at robot_localization doesn't look like it is able to be consumed either, correct me if I am wrong.

I'm happy to get any input which ideally would work for simulation, and the robot_localization. IDK if this means adding another format to robot_localization or something clunky like a flag to publish the same info under two different msg formats simultaneously.

**Comment by ahuer2435** [comment]
@mr337 
I have a question. RMC already contains heading likely. Why not directly publish heading in RMC?

**Comment by mr337** [comment]
@ahuer2435 Actually a good question. I did some digging and this is what I found and would be happy for input.

HDT is for a system that produces heading, or how a vehicle is pointing, not moving, just currently pointing. [1, 2] This is something that can only be produced by a system built for generating heading such as a dual antenna receiver.

RMC generates track angle [3,4] which is the angle I have been traveling over a set amount of time. This is not the same thing as heading since a vehicle can be pointed one direction but move another direction over time, like a boat that has a strong side wind.

1 - http://aprs.gids.nl/nmea/#hdt
2 - https://www.trimble.com/OEM_ReceiverHelp/V4.44/en/NMEA-0183messages_HDT.html
3 - https://www.trimble.com/OEM_ReceiverHelp/V4.44/en/NMEA-0183messages_RMC.html
4 - http://aprs.gids.nl/nmea/#rmc

**Comment by ahuer2435** [comment]
@mr337  Thank you very much for your detail answer. Should I understand as follows: 
For a car, if longitudinal movement，heading is  consistent with track angle.  But if lateral movement，they are different.
Am I right？

**Comment by mr337** [comment]
@ahuer2435 yes I believe you are spot on

File: package.xml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <package>
   <name>nmea_navsat_driver</name>
-  <version>0.5.0</version>
+  <version>0.5.1</version>
```
Comment (evenator) [inline package.xml]: Please revert the version number change. Version numbers only get bumped for releases.

File: package.xml [hunk 2]
```diff
@@ -25,6 +25,7 @@
   <build_depend>roslint</build_depend>
   <run_depend>rospy</run_depend>
   <run_depend>python-serial</run_depend>
+  <run_depend>geometry</run_depend>
```
Comment (evenator) [inline package.xml]: This package is not used. Instead, you should add `<run_depend>tf</run_depend>` for the new dependency on `tf`.

**Comment by evenator** [comment]
I retargeted this PR to master. Unfortunately, that introduced a conflict in the package.xml, so please merge `master` into this branch and fix that.

**Comment by mr337** [comment]
@evenator Thanks for the review. I can make all the recommended changes. Unfortunately I don't have access to a dual heading receiver at the time, though this may change in the future.

@naoki-mizuno Do you still have access to the Hemisphere ssV-102?

**Comment by naoki-mizuno** [comment]
@mr337 Yes I do. I can check it out on the device and let you know when the changes are ready.

**Comment by mr337** [comment]
Perfect, give me 2-3 days to get to this and I'll ping you when ready :)

**Comment by evenator** [comment]
Thanks all! I look forward to reviewing the final PR.

File: package.xml [hunk 3]
```diff
@@ -18,11 +18,12 @@
 
   <buildtool_depend>catkin</buildtool_depend>
   <build_depend>roslint</build_depend>
-  <exec_depend>rospy</exec_depend>
-  <exec_depend>python-serial</exec_depend>
-  <exec_depend>geometry_msgs</exec_depend>
-  <exec_depend>nmea_msgs</exec_depend>
-  <exec_depend>sensor_msgs</exec_depend>
+  <run_depend>rospy</run_depend>
```
Comment (naoki-mizuno) [inline package.xml]: I got an error saying `format="2"` shouldn't contain `run_depend`. The [doc](http://docs.ros.org/jade/api/catkin/html/howto/format2/migrating_from_format_1.html#run-depend) says `exec_depend` should be used.

**Comment by naoki-mizuno** [comment]
@mr337 Thanks for the heads-up! I'll check out the changes on the device within the next few days and get back to you.

File: package.xml (hunk 3 above)
Comment (evenator) [reply package.xml]: Sorry for the misdirection on that. I didn't realize this one was format 2

**Comment by naoki-mizuno** [comment]
After correcting the `run_depend`, I tested the code out on the ssV-102 and it was working without any errors. Although I wasn't sure how I can show from the data so I just plotted the heading (yaw) angle.
figure_1

50 - 80 s: going around in a CCW circle twice
80 - 90 s: going around in a CW circle once

If you need more evidence, I'd be glad to provide them!

**Comment by mr337** [comment]
I have updated the `package.xml`. I am still learning about ROS packaging so if anything else is off let me know and I'll be happy to fix it.

@naoki-mizuno Thanks for getting the real hardware test done so quickly!

**Comment by mr337** [comment]
I am open to either option and would be happy to build in Option 2. I don't think it would take too long, but still require naoki-mizuno to do another test run if available.

**Comment by evenator** [comment]
Can you resolve the merge conflicts? Thanks!

**Comment by mr337** [comment]
@evenator Oops this one fell through the cracks, I have rebased it. Unclear if it needs testing again before merging :S

**Comment by evenator** [comment]
Thanks for contributing! I've merged this PR to master, and it will be included in the release package for the upcoming ROS version (ROS N). As discussed above, this probably won't be backported to released versions of ROS.

**Comment by LSR20192872** [comment]
It's about increasing the heading. What is the final feature pack? Can you send me a link? thank you very much!

**Pull request merged**